### PR TITLE
Apply unpack_to_dest and use matmul_tiles instread of reduce_tile in w-dim

### PR DESCRIPTION
**Pull request by namhyung-kim-moreh**
### Ticket
https://github.com/tenstorrent/tt-metal/issues/13576

### Problem description
- Without `unpack_to_dest`, precision is lost when copying fp32 CB to dest register.
- `reduce_tiles` in w-dim has poor precision.

### What's changed
- Main changes
  - Apply `unpack_to_dest` to some moreh_ops.
  - Use `matmul_tiles` instread of `reduce_tiles` in w-dim.
- Miscellaneous changes
  - Do not check rank of tilized idx tensor in moreh_getitem.
  - Increase circular buffer size in moreh_matmul

### Checklist
- [x] Post commit CI passes
- [ ] Blackhole Post commit (if applicable)
- [ ] Model regression CI testing passes (if applicable)
- [ ] Device performance regression CI testing passes (if applicable)
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/moreh/moreh_getitem/device/moreh_getitem_device_operation.cpp [hunk 1]
```diff
@@ -34,8 +34,8 @@ void MorehGetItemOperation::validate_inputs(
         auto index_layout = index_tensor.get_layout();
         if (index_layout == Layout::ROW_MAJOR) {
             TT_FATAL(index_shape.rank() == 1, "Index tensor must be 1D for ROW_MAJOR layout!");
-        } else if (index_layout == Layout::TILE) {
-            TT_FATAL(index_shape.rank() == 5, "Index tensor must be 5D for TILE layout!");
+        } else {
```
Comment (dongjin-na) [inline ttnn/cpp/ttnn/operations/moreh/moreh_getitem/device/moreh_getitem_device_operation.cpp:37]: Would it be better to remove the empty else clause?

Comment (namhyeong-kim) [reply ttnn/cpp/ttnn/operations/moreh/moreh_getitem/device/moreh_getitem_device_operation.cpp:37]: Someone might be confused about whether omitting the `else` block was intentional or just a mistake, so I left the empty block.

**Review by dongjin-na — APPROVED** [review]

**Comment by dmakoviichuk-tt** [comment]
@dongjin-na I have a question:
in the moreh_common you have:

```
template <bool at_start, PoolType reduce_type = REDUCE_OP, ReduceDim reduce_dim = REDUCE_DIM>
ALWI void reduce_and_recip_tile_to_cb(
    uint32_t icb0, uint32_t icb1, uint32_t ocb, uint32_t size, uint32_t pop0 = 1, uint32_t pop1 = 1) {
    constexpr uint32_t onetile = 1;
    constexpr int dst0 = 0;

    cb_reserve_back(ocb, onetile);
    cb_wait_front(icb1, onetile);

    tile_regs_acquire();
    reduce_init_delta_with_dt<at_start, reduce_type, reduce_dim>(ocb, icb0, icb1);
    for (uint32_t x = 0; x < size; ++x) {
        cb_wait_front(icb0, x + 1);  // must be a cumulative wait for correctness

        constexpr uint32_t bcast_scaler0 = 0;  // 0th index from bcast_scaler CB
        reduce_tile<reduce_type, reduce_dim>(icb0, icb1, x, bcast_scaler0, dst0);
    }
    if (pop0) {
        cb_pop_front(icb0, pop0);
    }
    if (pop1) {
        cb_pop_front(icb1, pop1);
    }

    reduce_revert_delta(ocb);

    recip_tile_init();
    recip_tile(dst0);
    tile_regs_commit();

    tile_regs_wait();
    pack_tile_with_dt(dst0, ocb);
    tile_regs_release();

    cb_push_back(ocb, onetile);
}
```
Should you apply the same trick here for softmax?

**Comment by dongjin-na** [comment]
> @dongjin-na I have a question: in the moreh_common you have some reduce functions:
> 
> ```
> template <bool at_start, PoolType reduce_type = REDUCE_OP, ReduceDim reduce_dim = REDUCE_DIM>
> ALWI void reduce_and_recip_tile_to_cb(
>     uint32_t icb0, uint32_t icb1, uint32_t ocb, uint32_t size, uint32_t pop0 = 1, uint32_t pop1 = 1) {
>     constexpr uint32_t onetile = 1;
>     constexpr int dst0 = 0;
> 
>     cb_reserve_back(ocb, onetile);
>     cb_wait_front(icb1, onetile);
> 
>     tile_regs_acquire();
>     reduce_init_delta_with_dt<at_start, reduce_type, reduce_dim>(ocb, icb0, icb1);
>     for (uint32_t x = 0; x < size; ++x) {
>         cb_wait_front(icb0, x + 1);  // must be a cumulative wait for correctness
> 
>         constexpr uint32_t bcast_scaler0 = 0;  // 0th index from bcast_scaler CB
>         reduce_tile<reduce_type, reduce_dim>(icb0, icb1, x, bcast_scaler0, dst0);
>     }
>     if (pop0) {
>         cb_pop_front(icb0, pop0);
>     }
>     if (pop1) {
>         cb_pop_front(icb1, pop1);
>     }
> 
>     reduce_revert_delta(ocb);
> 
>     recip_tile_init();
>     recip_tile(dst0);
>     tile_regs_commit();
> 
>     tile_regs_wait();
>     pack_tile_with_dt(dst0, ocb);
>     tile_regs_release();
> 
>     cb_push_back(ocb, onetile);
> }
> ```
> 
> Should you apply the same trick here for softmax?

Hi, it should be okay to apply it as long as the same conditions are met (w-dim and unpack_to_dest are used).

**Pull request merged**